In [2]:
import yfinance as yf
from scipy.optimize import brentq
import datetime as dt
import numpy as np
import scipy.stats as stats

In [20]:
ticker=yf.Ticker('GOOG')
prices=ticker.history(period="2y")['Close']
expiry=ticker.options[2]
options=ticker.option_chain(expiry).calls 
last_price=prices.iloc[-1]
s0=last_price
margin=10
options=options[(options['strike']>=last_price-margin) & (options['strike']<=last_price+margin)]
mu=prices.pct_change().mean()
t=(dt.datetime.strptime(expiry,"%Y-%m-%d")-dt.datetime.today()).days+1
r=(1+0.049)**(1/255)-1
for k, l in zip(options['strike'],options['lastPrice']):
    def BS(sigma):
        d1=(np.log(s0/k)+(r+sigma**2/2)*t)/ (sigma*np.sqrt(t))
        d2=d1-sigma*np.sqrt(t)
        c=s0*stats.norm.cdf(d1)-k*np.exp(-r*t)*stats.norm.cdf(d2)
        return c
    def objective(sigma):
        return BS(sigma)
    implied_vol= brentq(objective,-0.01,0.07,xtol=1e-6)
    print(f"la volatilidad implicita para strike en {k} es: {implied_vol:.6f}")

print(f" precio pasado: {last_price}, tiempo:{t}, media:{mu} ")

la volatilidad implicita para strike en 245.0 es: -0.000053
la volatilidad implicita para strike en 250.0 es: -0.000016
la volatilidad implicita para strike en 255.0 es: 0.000001
la volatilidad implicita para strike en 260.0 es: 0.000043
 precio pasado: 253.72999572753906, tiempo:15, media:0.0013766286699059664 


In [3]:
ticker=yf.Ticker('GOOG')
prices=ticker.history(period="2y")['Close']
expiry=ticker.options[2]
options=ticker.option_chain(expiry).calls 
last_price=prices.iloc[-1]
s0=last_price
margin=10
options=options[(options['strike']>=last_price-margin) & (options['strike']<=last_price+margin)]
mu=prices.pct_change().mean()
t=(dt.datetime.strptime(expiry,"%Y-%m-%d")-dt.datetime.today()).days+1
r=(1+0.049)**(1/255)-1
for k, l in zip(options['strike'],options['lastPrice']):
    def BS(sigma):
        d1=(np.log(s0/k)+(r+sigma**2/2)*t)/ (sigma*np.sqrt(t))
        d2=d1-sigma*np.sqrt(t)
        c=s0*stats.norm.cdf(d1)-k*np.exp(-r*t)*stats.norm.cdf(d2)
        return c
    def objective(sigma):
        return BS(sigma)
    implied_vol= brentq(objective,-0.01,0.07,xtol=1e-6)
    print(f"la volatilidad implicita para strike en {k} es: {implied_vol:.6f}")

print(f" precio pasado: {last_price}, tiempo:{t}, media:{mu} ")

la volatilidad implicita para strike en 270.0 es: -0.000087
la volatilidad implicita para strike en 272.5 es: -0.000078
la volatilidad implicita para strike en 275.0 es: -0.000039
la volatilidad implicita para strike en 277.5 es: -0.000010
la volatilidad implicita para strike en 280.0 es: 0.000028
la volatilidad implicita para strike en 282.5 es: 0.000032
la volatilidad implicita para strike en 285.0 es: 0.000000
la volatilidad implicita para strike en 287.5 es: 0.000147
 precio pasado: 277.7900085449219, tiempo:17, media:0.0016805762648385116 


In [7]:

fecha_inicio = '2025-10-23'
fecha_fin = '2025-11-07' 

precios_rv = yf.download('GOOG', start=fecha_inicio, end=fecha_fin)['Close']
rendimientos = np.log(precios_rv / precios_rv.shift(1)).dropna()
sigma_diaria = rendimientos.std()

dias_habiles_año = 252
volatilidad_realizada = sigma_diaria * np.sqrt(dias_habiles_año)
volatilidad_realizada

C:\Users\marce\AppData\Local\Temp\ipykernel_10788\1778391811.py:4: FutureWarning: YF.download() has changed argument auto_adjust default to True
  precios_rv = yf.download('GOOG', start=fecha_inicio, end=fecha_fin)['Close']
[*********************100%***********************]  1 of 1 completed


Ticker
GOOG    0.312665
dtype: float64

En nuestro caso la accion de google subio bastante por lo que la volatilidad implicita no se calculo de un strike price tan alto, si  tuviera que decir una volatilidad implicita que pudiera acercarse a este caso podria ser alrededor del 170 o 190% ya que este volatilidad es anual pero las opciones se cortan cada semana por que para que en una accion suba 13 dolares en dos semanas es una volatilidad enorme. 

la volatilidad real fue del 31% que es altisima en el periodo de la opcion que elegimos nosotros.